README.md
# JVM Sounds

This is a very simple JVM library that plays JVM events as audio sounds.  
 
* Memory allocation rate as a sine tone
* Minor GC is a wood percussion
* Major GC is a noise hit 
* Hiccups (underlying JVM timejumps) are played as sample hiccups.

Please see [Logging vs Memory](https://tersesystems.com/blog/2020/07/09/logging-vs-memory/) and [the tweet](https://twitter.com/will_sargent/status/1281718634289573890).

## Examples

There are MP3 recordings available that show how the JVM sounds [using an example program](https://github.com/tersesystems/jvmsounds/blob/master/src/test/java/example/Main.java#L22):

* [Light Load with 10000 wastage](https://github.com/tersesystems/jvmsounds/blob/master/recordings/10000.mp3)
* [Medium Load with 20000 wastage](https://github.com/tersesystems/jvmsounds/blob/master/recordings/20000.mp3)
* [Heavy Load with 30000 wastage](https://github.com/tersesystems/jvmsounds/blob/master/recordings/30000.mp3)

There's also an example of a JVM running under Virtualbox while under a Gatling load test, which makes it prone to hiccups:

* [Gatling with Hiccups](https://github.com/tersesystems/jvmsounds/blob/master/recordings/gatling-with-hiccup.mp3)

## Resolver

The libraries are in Bintray under `tersesystems/maven`.  Installation instructions are similar to [terse-logback installation](https://tersesystems.github.io/terse-logback/installation/):

```groovy
repositories {
    maven {
        url  "https://dl.bintray.com/tersesystems/maven" 
    }
}
```

or SBT:

```
resolvers += Resolver.bintrayRepo("tersesystems", "maven")
```

## Library

The library has dependencies on jSyn and on `jvm-alloc-rate-meter` and `jvm-hiccup-meter`.  It is plain Java, no Scala library required.

```scala
libraryDependencies += "com.tersesystems.jvmsounds" % "jvmsounds" % "0.0.1"
```

## Java Agent

You can run this as a java agent.  I like using [sbt-javaagent](https://github.com/sbt/sbt-javaagent).

```scala
javaAgents += "com.tersesystems.jvmsounds" % "jvmsounds" % "0.0.1"
```

## Options

* `-n napTimeMs`
* `-at allocThresholdMs` - default is 300 MB/second, below that it is silent
* `-av allocVolume 0.0 - 1.0` - volume of the allocation tone, default 0.1
* `-mv minorGCVolume 0.0 - 1.0` - volume of the minor GC event, default 0.6
* `-Mv mixedGCVolume 0.0 - 1.0` - volume of the mixed GC event, default 0.6
* `-ht hiccupThresholdMs` - default is 50 ms, below that no hiccup
* `-hv hiccupVolume 0.0 - 1.0` - volume of the hiccup, default 1.0


docs/release-notes.md
<sup><sup>*Release notes were automatically generated by [Shipkit](http://shipkit.org/)*</sup></sup>

#### 0.0.1
 - 2020-07-19 - 20 commits by [Will Sargent](https://github.com/wsargent) - published to [![Bintray](https://img.shields.io/badge/Bintray-0.0.1-green.svg)](https://bintray.com/tersesystems/maven/jvmsounds/0.0.1)
 - No pull requests referenced in commit messages.



src/test/java/example/Main.java
/**
 * License
 * -------
 * Written in 2020 by Will Sargent will@tersesystems.com
 * To the extent possible under law, the author(s) have dedicated all copyright and related and neighboring rights to this software to the public domain worldwide. This software is distributed without any warranty.
 * You should have received a copy of the CC0 Public Domain Dedication along with this software. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>
 */
package example;

import com.tersesystems.jvmsounds.JVMSounds;

import java.io.BufferedWriter;
import java.io.FileWriter;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;

public class Main {

    public static void main(String[] args) throws InterruptedException {
        // Increase this number to waste more memory
        int wastage = 10000;

        try {
            JVMSounds jvmSounds = new JVMSounds();
            try {
                jvmSounds.start();
                createGarbage(wastage);
                Thread.sleep(7500 * 10);
            } finally {
                jvmSounds.stop();
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    // Create a very inefficient way of writing to a file.
    // I could probably just create a heap buffer and throw it away, but this is more "logging" oriented.
    private static void createGarbage(int wastage) {
        Runnable garbageProducingRunnable = () -> {
            try {
                Path tempFile = Files.createTempFile("tmp", "tmp");
                while (true) {
                    try (BufferedWriter writer = new BufferedWriter(new FileWriter(tempFile.toFile()), wastage)) {
                        long l = System.currentTimeMillis();
                        StringBuffer b = new StringBuffer().append(Math.random()).append("derp").append(l);
                        writer.write(b.toString());
                        if (Files.size(tempFile) > 1024 * 1000) {
                            Files.delete(tempFile);
                        }
                    }
                }
            } catch (IOException e) {
                e.printStackTrace();
            }
        };
        Thread t1 = new Thread(garbageProducingRunnable);
        t1.setDaemon(true);
        t1.start();
    }

}


src/test/java/example/GarbageCollectionPrinter.java
/**
 * License
 * -------
 * Written in 2020 by Will Sargent will@tersesystems.com
 * To the extent possible under law, the author(s) have dedicated all copyright and related and neighboring rights to this software to the public domain worldwide. This software is distributed without any warranty.
 * You should have received a copy of the CC0 Public Domain Dedication along with this software. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>
 */
package example;

import com.sun.management.GarbageCollectionNotificationInfo;

import java.lang.management.MemoryUsage;
import java.util.Map;

public class GarbageCollectionPrinter {

    public void printGarbageCollectionInfo(GarbageCollectionNotificationInfo info) {
        //get all the info and pretty print it
        long duration = info.getGcInfo().getDuration();
        String gctype = info.getGcAction();
        if ("end of minor GC".equals(gctype)) {
            gctype = "Young Gen GC";
        } else if ("end of major GC".equals(gctype)) {
            gctype = "Old Gen GC";
        }
        printGcType(info, duration, gctype);

        prettyPrintMemorySpace(info);
    }

    public void printGcType(GarbageCollectionNotificationInfo info, long duration, String gctype) {
        System.out.println();
        System.out.println(gctype + ": - " +
                info.getGcInfo().getId() +
                " " +
                info.getGcName() +
                " (from " +
                info.getGcCause() +
                ") " +
                duration +
                " milliseconds; start-end times " +
                info.getGcInfo().getStartTime() +
                "-" +
                info.getGcInfo().getEndTime());
    }

    public void prettyPrintMemorySpace(GarbageCollectionNotificationInfo info) {
        //Get the information about each memory space, and pretty print it
        Map<String, MemoryUsage> membefore = info.getGcInfo().getMemoryUsageBeforeGc();
        Map<String, MemoryUsage> mem = info.getGcInfo().getMemoryUsageAfterGc();
        for (Map.Entry<String, MemoryUsage> entry : mem.entrySet()) {
            String name = entry.getKey();
            MemoryUsage memdetail = entry.getValue();
            long memInit = memdetail.getInit();
            long memCommitted = memdetail.getCommitted();
            long memMax = memdetail.getMax();
            long memUsed = memdetail.getUsed();
            MemoryUsage before = membefore.get(name);
            long used = before.getUsed();
            if (before.getCommitted() > 0) {
                long beforepercent = (used > 0) ? ((used * 1000L) / before.getCommitted()) : 0;
                long percent = (memUsed > 0) ? ((memUsed * 1000L) / before.getCommitted()) : 0; //>100% when it gets expanded
                printMemorySpace(name, memCommitted, memMax, memUsed, beforepercent, percent);
            }
        }
        System.out.println();
        long totalGcDuration = info.getGcInfo().getDuration();
        long percent = totalGcDuration * 1000L / info.getGcInfo().getEndTime();
        System.out.println("GC cumulated overhead " + (percent / 10) + "." + (percent % 10) + "%");
    }

    public void printMemorySpace(String name, long memCommitted, long memMax, long memUsed, long beforepercent, long percent) {
        String s1 = memCommitted == memMax ? "(fully expanded)" : "(still expandable)";
        String s3 = "used: " +
                (beforepercent / 10) +
                "." +
                (beforepercent % 10) +
                "%->" +
                (percent / 10) +
                "." +
                (percent % 10) +
                "%(" +
                ((memUsed / 1048576) + 1) +
                "MB) / ";
        String s = name + s1 + s3;
        System.out.print(s);
    }
}


src/main/java/com/tersesystems/alloctone/SoftSynth.java
package com.tersesystems.alloctone;

import javax.sound.sampled.AudioFormat;
import javax.sound.sampled.AudioSystem;
import javax.sound.sampled.DataLine;
import javax.sound.sampled.SourceDataLine;
import java.time.Duration;
import java.util.Arrays;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;
import java.util.function.DoubleSupplier;

/**
 * Plays a continously changing tone from a supplier.
 *
 * The sample player code is taken from the
 * <a href="https://www.drdobbs.com/jvm/music-components-in-java-creating-oscill/230500178">Music Components in Java: Creating Oscillators</a>
 * article in Dr Dobbs, written by Craig Lindley.
 */
public class SoftSynth {
    private final ScheduledExecutorService st = Executors.newScheduledThreadPool(1, r -> {
        Thread t = new Thread(r);
        t.setName("alloctone-scheduled-daemon");
        t.setDaemon(true);
        return t;
    });

    private final SamplePlayer player = new SamplePlayer();

    public SoftSynth(DoubleSupplier supplier) {
        this(supplier, Duration.ofMillis(50));
    }

    public SoftSynth(DoubleSupplier supplier, Duration interval) {
        BasicOscillator osc = new BasicOscillator();
        osc.setWaveshape(BasicOscillator.WAVESHAPE.SIN);

        Runnable runnable = () -> {
            double frequency = supplier.getAsDouble();
            osc.setFrequency(frequency);
        };
        start(osc, runnable, interval);
    }

    void start(ISampleProvider sampleProvider, Runnable runnable, Duration interval) {
        st.scheduleAtFixedRate(runnable, 0, interval.toMillis(), TimeUnit.MILLISECONDS);
        player.setSampleProvider(sampleProvider);
        player.startPlayer();
    }

    public void stop() {
        st.shutdown();
        player.stopPlayer();
    }

    public interface ISampleProvider {
        int getSamples(byte[] buffer);
    }

    public static class BasicOscillator implements ISampleProvider {
        private BasicOscillator.WAVESHAPE waveshape;
        private long periodSamples;
        private long sampleNumber;

        public enum WAVESHAPE {
            SIN, SQU, SAW
        }

        public BasicOscillator() {
            setWaveshape(BasicOscillator.WAVESHAPE.SIN);
            setFrequency(1000.0);
        }

        public void setWaveshape(BasicOscillator.WAVESHAPE waveshape) {
            this.waveshape = waveshape;
        }

        public void setFrequency(double frequency) {
            periodSamples = (long) (SamplePlayer.SAMPLE_RATE / frequency);
        }

        protected double getSample() {
            double value;
            double x = sampleNumber / (double) periodSamples;

            switch (waveshape) {
                default:

                case SIN:
                    value = Math.sin(2.0 * Math.PI * x);
                    break;

                case SQU:
                    if (sampleNumber < (periodSamples / 2)) {
                        value = 1.0;
                    } else {
                        value = -1.0;
                    }
                    break;

                case SAW:
                    value = 2.0 * (x - Math.floor(x + 0.5));
                    break;
            }
            sampleNumber = (sampleNumber + 1) % periodSamples;
            return value;
        }

        public int getSamples(byte[] buffer) {
            int index = 0;
            for (int i = 0; i < SamplePlayer.SAMPLES_PER_BUFFER; i++) {
                double ds = getSample() * Short.MAX_VALUE;
                short ss = (short) Math.round(ds);
                buffer[index++] = (byte) (ss >> 8);
                buffer[index++] = (byte) (ss & 0xFF);
            }
            return SamplePlayer.BUFFER_SIZE;
        }
    }

    public static class SamplePlayer extends Thread implements ISampleProvider {
        // Count of zeroed buffers to return before switching to real sample provider
        private static final int TEMP_BUFFER_COUNT = 20;

        // AudioFormat parameters
        public static final int SAMPLE_RATE = 22050;
        private static final int SAMPLE_SIZE = 16;
        private static final int CHANNELS = 1;
        private static final boolean SIGNED = true;
        private static final boolean BIG_ENDIAN = true;

        // Chunk of audio processed at one time
        public static final int BUFFER_SIZE = 1000;
        public static final int SAMPLES_PER_BUFFER = BUFFER_SIZE / 2;

        // Instance data
        private final AudioFormat format;
        private final DataLine.Info info;
        private SourceDataLine auline;
        private boolean hasRun;
        private boolean done;
        private int bufferCount;
        private final byte[] sampleData = new byte[BUFFER_SIZE];
        private ISampleProvider provider;
        private ISampleProvider realProvider;

        public SamplePlayer() {
            format = new AudioFormat(SAMPLE_RATE, SAMPLE_SIZE, CHANNELS, SIGNED, BIG_ENDIAN);
            info = new DataLine.Info(SourceDataLine.class, format);
            Arrays.fill(sampleData, (byte) 0);
            // Set temp provider so zeroed buffers are consumed initially
            provider = this;
        }

        /**
         * Process a buffer full of samples pulled from the sample provider
         * <p>
         * NOTE: this class acts as a sample provider to itself during<br>
         * the initialization of the sound engine. Once bufferCount of buffers<br>
         * has been processed, SamplePlayer switches to its real sample provider.<br>
         * This is necessary to prevent glitches on startup.
         *
         * @param buffer Buffer in which the samples are to be processed
         * @return Count of number of bytes processed
         */
        public int getSamples(byte[] buffer) {
            bufferCount++;
            if (bufferCount >= TEMP_BUFFER_COUNT) {
                // Audio system flushed so switch to real sample provider
                provider = realProvider;
            }
            return BUFFER_SIZE;
        }

        public void run() {
            done = false;
            int nBytesRead = 0;
            try {
                // Get line to write data to
                auline = (SourceDataLine) AudioSystem.getLine(info);
                auline.open(format);
                auline.start();

                while ((nBytesRead != -1) && (!done)) {
                    nBytesRead = provider.getSamples(sampleData);
                    if (nBytesRead > 0) {
                        auline.write(sampleData, 0, nBytesRead);
                    }
                }
            } catch (Exception e) {
                e.printStackTrace();
            } finally {
                auline.drain();
                auline.close();
            }
        }

        public void startPlayer() {
            if (hasRun) {
                System.err.println("Illegal to restart a Thread once it has been stopped");
                return;
            }
            if (realProvider != null) {
                hasRun = true;
                start();
            }
        }

        public void stopPlayer() {
            done = true;
        }

        public void setSampleProvider(ISampleProvider provider) {
            realProvider = provider;
        }
    }
}


src/main/java/com/tersesystems/jvmsounds/JVMSounds.java
/**
 * License
 * -------
 * Written in 2020 by Will Sargent will@tersesystems.com
 * To the extent possible under law, the author(s) have dedicated all copyright and related and neighboring rights to this software to the public domain worldwide. This software is distributed without any warranty.
 * You should have received a copy of the CC0 Public Domain Dedication along with this software. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>
 */
package com.tersesystems.jvmsounds;

import com.jsyn.JSyn;
import com.jsyn.Synthesizer;
import com.jsyn.data.AudioSample;
import com.jsyn.instruments.DrumWoodFM;
import com.jsyn.instruments.NoiseHit;
import com.jsyn.unitgen.*;
import com.jsyn.util.SampleLoader;
import com.sun.management.GarbageCollectionNotificationInfo;
import com.sun.management.GcInfo;
import jvm_alloc_rate_meter.MeterThread;

import javax.management.ListenerNotFoundException;
import javax.management.Notification;
import javax.management.NotificationEmitter;
import javax.management.NotificationListener;
import javax.management.openmbean.CompositeData;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.InputStream;
import java.lang.instrument.Instrumentation;
import java.lang.management.GarbageCollectorMXBean;
import java.lang.management.ManagementFactory;
import java.util.List;
import java.util.function.Consumer;
import java.util.logging.Level;
import java.util.logging.Logger;

import static com.sun.management.GarbageCollectionNotificationInfo.GARBAGE_COLLECTION_NOTIFICATION;
import static com.sun.management.GarbageCollectionNotificationInfo.from;

/**
 * A class which plays the memory allocation tone rate as a sine tone through
 * <a href="https://github.com/philburk/jsyn">jSyn</a>.
 * <p>
 * Please see the <a href="https://tersesystems.com/blog/2020/07/09/logging-vs-memory/">Logging vs Memory</a> for more details.
 * <p>
 * The allocation rate is taken from <a href="https://github.com/clojure-goes-fast/jvm-alloc-rate-meter">jvm-alloc-rate-meter</a> by
 * alexander-yakushev.
 */
public class JVMSounds {

    /** Default allocation threshold in megabytes per second. */
    public static final int DEFAULT_ALLOC_THRESHOLD = 300;

    public static final double DEFAULT_ALLOC_VOLUME = 0.1;

    public static final double DEFAULT_MINOR_GC_VOLUME = 0.6;

    public static final double DEFAULT_MIXED_GC_VOLUME = 0.6;

    /** Default hiccup threshold in milliseconds. */
    public static final int DEFAULT_HICCUP_THRESHOLD = 50;
    public static final double DEFAULT_HICCUP_VOLUME = 1.0;
    public static final int DEFAULT_NAP_TIME = 0;

    private final Logger logger = Logger.getLogger(JVMSounds.class.getName());

    private final Synthesizer synth;
    private final LineOut lineOut;

    private final SineOscillator allocTone;
    private final DrumWoodFM minorGcSound;
    private final NoiseHit mixedGcSound;
    private final AudioSample hiccupSound;
    private final VariableRateDataReader hiccupPlayer;

    private final HiccupProducer hiccupProducer;
    private final AllocationRateProducer allocProducer;
    private final GCEventProducer gcEventProducer;

    private final Configuration config = new Configuration();

    public static void premain(String argsString, Instrumentation inst) throws IOException {
        String[] args = (argsString != null) ? argsString.split("[ ,;]+") : new String[0];
        commonMain(args);
    }

    private static void nap(long millis) {
        try {
            long startMillis = System.currentTimeMillis();
            while (System.currentTimeMillis() - startMillis < millis) {
                Thread.sleep(10);
            }
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
    }

    public static JVMSounds commonMain(String[] args) throws IOException {
        JVMSounds jvmSounds = null;
        try {
            jvmSounds = new JVMSounds(args);
            jvmSounds.start();
        } catch (FileNotFoundException e) {
            System.err.println("heaputils: Failed to open log file.");
        }
        return jvmSounds;
    }

    public JVMSounds() throws IOException {
        this(new String[0]);
    }

    public JVMSounds(String[] args) throws IOException {
        config.parseArgs(args);

        synth = JSyn.createSynthesizer();
        allocTone = new SineOscillator();
        lineOut = new LineOut();
        minorGcSound = new DrumWoodFM();
        mixedGcSound = new NoiseHit();

        synth.add(allocTone);
        synth.add(lineOut);
        synth.add(minorGcSound);
        synth.add(mixedGcSound);

        allocTone.output.connect(0, lineOut.input, 0);
        allocTone.output.connect(0, lineOut.input, 1);

        minorGcSound.getOutput().connect(0, lineOut.input, 0);
        minorGcSound.getOutput().connect(0, lineOut.input, 1);

        mixedGcSound.getOutput().connect(0, lineOut.input, 0);
        mixedGcSound.getOutput().connect(0, lineOut.input, 1);

        allocProducer = new AllocationRateProducer(this::setAllocTone);
        gcEventProducer = new GCEventProducer(this::playMinorGC, this::playMixedGC);
        hiccupProducer = new HiccupProducer(this::hiccup);
        SampleLoader.setJavaSoundPreferred(false);
        InputStream resourceAsStream = this.getClass().getClassLoader().getResourceAsStream("hiccup.wav");
        hiccupSound = SampleLoader.loadFloatSample(resourceAsStream);

        if (hiccupSound.getChannelsPerFrame() == 1) {
            synth.add(hiccupPlayer = new VariableRateMonoReader());
            hiccupPlayer.output.connect(0, lineOut.input, 0);
        } else if (hiccupSound.getChannelsPerFrame() == 2) {
            synth.add(hiccupPlayer = new VariableRateStereoReader());
            hiccupPlayer.output.connect(0, lineOut.input, 0);
            hiccupPlayer.output.connect(1, lineOut.input, 1);
        } else {
            throw new RuntimeException("Can only play mono or stereo samples.");
        }
        hiccupPlayer.rate.set(hiccupSound.getFrameRate());

        if (config.napTime > 0) {
            nap(config.napTime);
        }
    }

    public void playMinorGC(GcInfo info) {
        logger.log(Level.FINE, "Minor GC {0}", info);
        minorGcSound.noteOn(300, config.minorGcVolume, synth.createTimeStamp());
    }

    public void playMixedGC(GcInfo info) {
        logger.log(Level.FINE, "Mixed GC {0}", info);
        mixedGcSound.noteOn(200, config.mixedGcVolume, synth.createTimeStamp());
    }

    public void hiccup(long hiccupNs) {
        double millis = hiccupNs / 1e6;
        logger.log(Level.FINE, "Hiccup duration in millis = {0}", millis);
        if (millis > config.hiccupThreshold) {
            hiccupPlayer.dataQueue.queue(hiccupSound);
        }
    }

    public void setAllocTone(double frequency) {
        // https://github.com/philburk/jsyn/blob/master/tests/com/jsyn/examples/ChebyshevSong.java
        // XXX ideally should set this to something that maps it to the closest note
        logger.log(Level.FINE, "Allocation rate = {0} MB/sec", frequency);
        if (frequency < config.allocThreshold) {
            allocTone.amplitude.set(0);
        } else {
            allocTone.frequency.set(frequency);
            allocTone.amplitude.set(config.allocVolume);
        }
    }

    public void start() {
        synth.start();
        allocTone.start();
        lineOut.start();

        hiccupProducer.start();
        allocProducer.start();
        gcEventProducer.start();

        hiccupPlayer.amplitude.set(config.hiccupVolume);
        allocTone.amplitude.set(config.allocVolume);
    }

    public void stop() {
        gcEventProducer.stop();
        allocProducer.stop();
        hiccupProducer.stop();
        lineOut.stop();
        allocTone.stop();
        synth.stop();
    }

    // https://www.fasterj.com/articles/gcnotifs.shtml
    private static class GCEventProducer {
        private final NotificationListener listener = new GCListener();
        private final Consumer<GcInfo> minorConsumer;
        private final Consumer<GcInfo> majorConsumer;

        public GCEventProducer(Consumer<GcInfo> minorConsumer, Consumer<GcInfo> majorConsumer) {
            this.minorConsumer = minorConsumer;
            this.majorConsumer = majorConsumer;
        }

        public void start() {
            List<GarbageCollectorMXBean> gcbeans = ManagementFactory.getGarbageCollectorMXBeans();
            for (GarbageCollectorMXBean gcbean : gcbeans) {
                NotificationEmitter emitter = (NotificationEmitter) gcbean;
                emitter.addNotificationListener(listener, null, null);
            }
        }

        public void stop() {
            List<GarbageCollectorMXBean> gcbeans = ManagementFactory.getGarbageCollectorMXBeans();
            for (GarbageCollectorMXBean gcbean : gcbeans) {
                NotificationEmitter emitter = (NotificationEmitter) gcbean;
                try {
                    emitter.removeNotificationListener(listener);
                } catch (ListenerNotFoundException e) {
                    // don't care
                }
            }
        }

        class GCListener implements NotificationListener {
            @Override
            public void handleNotification(Notification notification, Object handback) {
                if (notification.getType().equals(GARBAGE_COLLECTION_NOTIFICATION)) {
                    GarbageCollectionNotificationInfo info = from((CompositeData) notification.getUserData());
                    String gctype = info.getGcAction();
                    if ("end of minor GC".equals(gctype)) {
                        minorConsumer.accept(info.getGcInfo());
                    } else {
                        majorConsumer.accept(info.getGcInfo());
                    }
                }
            }
        }
    }

    // https://github.com/clojure-goes-fast/jvm-alloc-rate-meter
    private static class AllocationRateProducer {
        private final MeterThread meterThread;

        public AllocationRateProducer(Consumer<Double> rateConsumer) {
            int intervalMs = 50;
            meterThread = new MeterThread(bytes -> {
                double mbytes = bytes / 1e6;
                rateConsumer.accept(mbytes);
            }, intervalMs);
        }

        public void start() {
            meterThread.start();
        }

        public void stop() {
            meterThread.terminate();
        }
    }

    // https://github.com/clojure-goes-fast/jvm-hiccup-meter
    private static class HiccupProducer {
        private final jvm_hiccup_meter.MeterThread meterThread;

        public HiccupProducer(Consumer<Long> callback) {
            int resolutionMs = 10;
            meterThread = new jvm_hiccup_meter.MeterThread(callback::accept, resolutionMs);
        }

        public void start() {
            meterThread.start();
        }

        public void stop() {
            meterThread.terminate();
        }
    }

    static class Configuration {
        public long napTime = DEFAULT_NAP_TIME;

        public double allocThreshold = DEFAULT_ALLOC_THRESHOLD;
        public double allocVolume = DEFAULT_ALLOC_VOLUME;

        public double minorGcVolume = DEFAULT_MINOR_GC_VOLUME;
        public double mixedGcVolume = DEFAULT_MIXED_GC_VOLUME;

        public long hiccupThreshold = DEFAULT_HICCUP_THRESHOLD;
        public double hiccupVolume = DEFAULT_HICCUP_VOLUME;

        public void parseArgs(String[] args) {
            for (int i = 0; i < args.length; ++i) {
                if (args[i].equals("-n")) {
                    napTime = Long.parseLong(args[++i]);
                } else if (args[i].equals("-at")) {
                    allocThreshold = Double.parseDouble(args[++i]);
                } else if (args[i].equals("-av")) {
                    allocVolume = Double.parseDouble(args[++i]);
                } else if (args[i].equals("-mv")) {
                    minorGcVolume = Double.parseDouble(args[++i]);
                }else if (args[i].equals("-Mv")) {
                    mixedGcVolume = Double.parseDouble(args[++i]);
                } else if (args[i].equals("-ht")) {
                    hiccupThreshold = Long.parseLong(args[++i]);
                } else if (args[i].equals("-hv")) {
                     hiccupVolume = Double.parseDouble(args[++i]);
                 } else {
                    System.out.println("Usage: " +
                            "[-n napTimeMs] " +
                            "[-at allocThresholdMs] " +
                            "[-av allocVolume 0.0 - 1.0] " +
                            "[-mv minorGCVolume 0.0 - 1.0] " +
                            "[-Mv mixedGCVolume 0.0 - 1.0] " +
                            "[-ht hiccupThresholdMs] " +
                            "[-hv hiccupVolume 0.0 - 1.0] ");

                    System.exit(1);
                }
            }
        }
    }

}


src/main/java/com/tersesystems/alloctone/Main.java
/*
 * SPDX-License-Identifier: CC0-1.0
 *
 * Copyright 2020 Will Sargent.
 *
 * Licensed under the CC0 Public Domain Dedication;
 * You may obtain a copy of the License at
 *
 *  http://creativecommons.org/publicdomain/zero/1.0/
 */
package com.tersesystems.alloctone;

import com.sun.management.GarbageCollectionNotificationInfo;
import jvm_alloc_rate_meter.MeterThread;

import javax.management.Notification;
import javax.management.NotificationEmitter;
import javax.management.NotificationListener;
import javax.management.openmbean.CompositeData;
import java.io.BufferedWriter;
import java.io.FileWriter;
import java.io.IOException;
import java.lang.management.GarbageCollectorMXBean;
import java.lang.management.MemoryUsage;
import java.nio.file.Files;
import java.nio.file.Path;
import java.time.Duration;
import java.util.List;
import java.util.Map;
import java.util.concurrent.atomic.AtomicLong;
import java.util.function.DoubleSupplier;

/**
 * A class which plays the memory allocation tone rate as a sine tone through
 * <a href="https://github.com/philburk/jsyn">jSyn</a>.
 * <p>
 * Please see the <a href="https://tersesystems.com/blog/2020/07/09/logging-vs-memory/">Logging vs Memory</a> for more details.
 * <p>
 * The allocation rate is taken from <a href="https://github.com/clojure-goes-fast/jvm-alloc-rate-meter">jvm-alloc-rate-meter</a> by
 * alexander-yakushev.
 */
public class Main {

    public static void main(String[] args) throws InterruptedException {
        // Increase this number to waste more memory
        int wastage = 1000;

        AllocationTone tone = new AllocationTone();
        try {
            AtomicLong allocRate = new AtomicLong(0L);
            DoubleSupplier supplier = () -> {
                double v = allocRate.longValue() / 1e6;
                System.out.println("Allocation rate in MiB/sec = " + v);
                return v;
            };
            tone.start(supplier, Duration.ofMillis(50));
            tone.setVolume(0.1);

            MeterThread meterThread = new MeterThread(allocRate::set);
            meterThread.start();

            List<GarbageCollectorMXBean> gcbeans = java.lang.management.ManagementFactory.getGarbageCollectorMXBeans();
            for (GarbageCollectorMXBean gcbean : gcbeans) {
                NotificationEmitter emitter = (NotificationEmitter) gcbean;
                NotificationListener listener = new GarbageCollectionListener();
                emitter.addNotificationListener(listener, null, null);
            }

            createGarbage(wastage);
            Thread.sleep(7500 * 1000);
        } finally {
            tone.stop();
        }
    }

    // Create a very inefficient way of writing to a file.
    private static void createGarbage(int wastage) {
        Runnable garbageProducingRunnable = () -> {
            try {
                Path tempFile = Files.createTempFile("tmp", "tmp");
                while (true) {
                    try (BufferedWriter writer = new BufferedWriter(new FileWriter(tempFile.toFile()), wastage)) {
                        long l = System.currentTimeMillis();
                        StringBuffer b = new StringBuffer().append(Math.random()).append("derp").append(l);
                        writer.write(b.toString());
                        if (Files.size(tempFile) > 1024 * 1000) {
                            Files.delete(tempFile);
                        }
                    }
                }
            } catch (IOException e) {
                e.printStackTrace();
            }
        };
        Thread t1 = new Thread(garbageProducingRunnable);
        t1.setDaemon(true);
        t1.start();
    }

    /**
     * Print out some GC statistics.  Maybe we can make record scratches and notes on GC events.
     */
    private static class GarbageCollectionListener implements NotificationListener {
        @Override
        public void handleNotification(Notification notification, Object handback) {
            if (notification.getType().equals(GarbageCollectionNotificationInfo.GARBAGE_COLLECTION_NOTIFICATION)) {
                //get the information associated with this notification
                GarbageCollectionNotificationInfo info = GarbageCollectionNotificationInfo.from((CompositeData) notification.getUserData());
                //get all the info and pretty print it
                long duration = info.getGcInfo().getDuration();
                String gctype = info.getGcAction();
                if ("end of minor GC".equals(gctype)) {
                    gctype = "Young Gen GC";
                } else if ("end of major GC".equals(gctype)) {
                    gctype = "Old Gen GC";
                }
                System.out.println();
                System.out.println(gctype + ": - " + info.getGcInfo().getId() + " " + info.getGcName() + " (from " + info.getGcCause() + ") " + duration + " milliseconds; start-end times " + info.getGcInfo().getStartTime() + "-" + info.getGcInfo().getEndTime());

                //Get the information about each memory space, and pretty print it
                Map<String, MemoryUsage> membefore = info.getGcInfo().getMemoryUsageBeforeGc();
                Map<String, MemoryUsage> mem = info.getGcInfo().getMemoryUsageAfterGc();
                for (Map.Entry<String, MemoryUsage> entry : mem.entrySet()) {
                    String name = entry.getKey();
                    MemoryUsage memdetail = entry.getValue();
                    long memInit = memdetail.getInit();
                    long memCommitted = memdetail.getCommitted();
                    long memMax = memdetail.getMax();
                    long memUsed = memdetail.getUsed();
                    MemoryUsage before = membefore.get(name);
                    long beforepercent = ((before.getUsed() * 1000L) / before.getCommitted());
                    long percent = ((memUsed * 1000L) / before.getCommitted()); //>100% when it gets expanded

                    System.out.print(name + (memCommitted == memMax ? "(fully expanded)" : "(still expandable)") + "used: " + (beforepercent / 10) + "." + (beforepercent % 10) + "%->" + (percent / 10) + "." + (percent % 10) + "%(" + ((memUsed / 1048576) + 1) + "MB) / ");
                }
                System.out.println();
                long totalGcDuration = info.getGcInfo().getDuration();
                long percent = totalGcDuration * 1000L / info.getGcInfo().getEndTime();
                System.out.println("GC cumulated overhead " + (percent / 10) + "." + (percent % 10) + "%");
            }
        }
    }
}
